hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/WritableKVStateBase.java

```diff
@@ -0,0 +1,265 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.spi.state;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.*;
+
+/**
+ * A base class for implementations of {@link WritableKVState}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public abstract class WritableKVStateBase<K extends Comparable<K>, V>
+        extends ReadableKVStateBase<K, V> implements WritableKVState<K, V> {
+    /** A map of all modified values buffered in this mutable state */
+    private final Map<K, V> modifications = new HashMap<>();
+
+    /**
+     * Create a new StateBase.
+     *
+     * @param stateKey The state key. Cannot be null.
+     */
+    protected WritableKVStateBase(@NonNull final String stateKey) {
+        super(stateKey);
+    }
+
+    /**
+     * Flushes all changes into the underlying data store. This method should <strong>ONLY</strong>
+     * be called by the code that created the {@link WritableKVStateBase} instance or owns it. Don't
+     * cast and commit unless you own the instance!
+     */
+    public final void commit() {
+        for (final var entry : modifications.entrySet()) {
+            final var key = entry.getKey();
+            final var value = entry.getValue();
+            if (value == null) {
+                removeFromDataSource(key);
+            } else {
+                putIntoDataSource(key, value);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Clears the set of modified keys and removed keys. Equivalent semantically to a "rollback"
+     * operation.
+     */
+    @Override
+    public final void reset() {
+        super.reset();
+        modifications.clear();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Nullable
+    public final V get(@NonNull K key) {
+        // If there is a modification, then we've already done a "put" or "remove"
+        // and should return based on the modification
+        if (modifications.containsKey(key)) {
+            final var value = modifications.get(key);
+            super.markRead(key, value);
+            return value;
+        } else {
+            return super.get(key);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Nullable
+    public final V getForModify(@NonNull final K key) {
+        Objects.requireNonNull(key);
+        // If there is a modification, then we've already done a "put" or "remove"
+        // and should return based on the modification
+        if (modifications.containsKey(key)) {
+            return modifications.get(key);
```

Comment:
Assuming that the key is contained in `modifications`, it would already have been marked as read. However, if it's possible that a modification can be made without actually reading the state beforehand (which would mark that object as read), it seems like `super.markRead` would also be needed here...

Reply:
Let my try to reconstruct the thinking.

First, we *do not* want to mark a modified key as a read key. Suppose I have some key `A_KEY` and I never read it, but I do `put` a new value for `A_KEY`. I don't care if somebody else in some other transaction changed the value of `A_KEY` because I *never* read it. I just put a new value. So I wouldn't want to invalidate my transaction's pre-handle work just because some other transaction modified `A_KEY`'s value as well.

In the same way, I also don't want to let 'readKeys' contain a key that I put but never read from the underlying data source, because I don't care what was in the underlying data source! I am just returning from get what I put, so if some previous transaction also modified the `A_KEY` value, I don't care! And I shouldn't invalidate the result.

So actually, I think this code is right, but the code in `WritableKVStateBase#get` is wrong and shouldn't update the read keys at all!!